odoo/models.py

```diff
@@ -5657,28 +5657,35 @@ def _cache(self):
         """ Return the cache of ``self``, mapping field names to values. """
         return RecordCache(self)
 
-    @api.model
     def _in_cache_without(self, field, limit=PREFETCH_MAX):
         """ Return records to prefetch that have no value in cache for ``field``
             (:class:`Field` instance), including ``self``.
             Return at most ``limit`` records.
         """
+        recs = self.browse(unique(self._prefetch_ids))
+        ids = self.env.cache.get_missing_ids(recs, field)
         # This method returns records that are either all real, or all new.
         # Those records are aimed at being either fetched, or computed.  But the
         # method '_fetch_field' is not correct with new records: it considers
         # them as forbidden records, and clears their cache!  On the other hand,
         # compute methods are not invoked with a mix of real and new records for
         # the sake of code simplicity.
+        return self._batch_with(ids, limit=limit)
+
+    def _batch_with(self, ids, limit=PREFETCH_MAX):
+        """ Return a batch of at most ``limit`` records from the record ``self``
+        and a subset of ``ids``.  The batch is such that its records are either
+        all real, or all new.
+        """
         kind = bool(self.id)
-        recs = self.browse(unique(self._prefetch_ids))
-        ids = [self.id]
-        for record_id in self.env.cache.get_missing_ids(recs - self, field):
-            if bool(record_id) != kind:
+        res_ids = OrderedSet([self.id])
+        for id_ in ids:
+            if bool(id_) != kind or id_ in res_ids:
```

Comment:
Does it really happen to have duplicate ids in `_prefetch_ids` / `env.all.to_compute` (is the `id_ in res_ids` needed) ?

Reply:
In the case of `env.all.to_compute`, it is needed for `self` (I may actually simplify the condition as `id_ != self.id`).
It is simpler and more efficient to deduplicate here than at the source.